### PR TITLE
Revise AlluxioEtcdClient connect logic

### PR DESCRIPTION
**Pull request by lucyge2022**
### What changes are proposed in this pull request?

fix bug where membership module close the AlluxioEtcdClient and wouldn't reconnect
lose connect logic, use one jetcd client throughout entire AlluxioEtcdClient lifecycle
introduce "nodes" command category to administrate worker membership

### Does this PR introduce any user facing changes?
N/A

### CODE REVIEW

**Comment by lucyge2022** [comment]
@yyongycy  this fixes(redid) the connect logic u pointed out in previous PR, plz take a look thanks!

File: dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java [hunk 1]
```diff
@@ -204,6 +204,6 @@ public void decommission(WorkerInfo worker) throws IOException {
 
   @Override
   public void close() throws Exception {
-    mAlluxioEtcdClient.close();
+    // NOTHING TO CLOSE
```
Comment (yyongycy) [inline dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java:207]: so will never close the client?

Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/membership/EtcdMembershipManager.java:207]: So AlluxioEtcdClient will be used as a singleton throughout entire process lifecycle, all reconnect/loadbalance logic is already handled by underlying jetcd client, I don't want to add such logic at our layer to complicate things as of now. And as MembershipManager is also used in client, such as FileSystemContext, which is constantly closed and reopen, it shouldn't be closing AlluxioEtcdClient as its shared with others.

File: dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java [hunk 2]
```diff
@@ -127,36 +126,41 @@ public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
...
+  private <V> V retryInternal(String description, RetryPolicy retryPolicy,
+                              EtcdUtilCallable<V> etcdCallable) throws IOException {
+    Exception ex = null;
+    // TODO(lucy) Currently retry on all sorts of exception and report the last exception,
+    // As jetcd exception often hides underneath CompletableFuture.get(), find
+    // more info later on distinguishing different possible exceptions.
+    while (retryPolicy.attempt()) {
+      try {
+        return etcdCallable.call();
+      } catch (Exception e) {
```
Comment (yyongycy) [inline dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:149]: retry for whatever reason?

What if it is the fatal error? eg, path not exist?

Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:149]: oh that's not fatal at all, etcd will not throw exception bcos certain path doesn't exist.
Currently EtcdException( the jetcd exception) is mostly for connection related, usually manifest itself in callback to those StreamObservers,  I haven't captured in inline call, we get more experience playing with jetcd and handle later

**Review by yyongycy — APPROVED** [review]
lgtm

File: dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java [hunk 3]
```diff
@@ -127,36 +124,41 @@ public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
...
+        // TODO(lucy) check in future if io.etcd.jetcd.common.exception.EtcdException
+        // needs to be handled for detection of underlying jetcd client connection
+        // getting into possible stale states or needs reconnection, for now, reuse single
+        // jetcd client handle for entire lifecycle of AlluxioEtcdClient.
+        LOG.warn("Failed to {} (attempt {}): {}", description, retryPolicy.getAttemptCount(),
```
Comment (jja725) [inline dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:152]: https://github.com/etcd-io/jetcd/blob/main/jetcd-common/src/main/java/io/etcd/jetcd/common/exception/EtcdExceptionFactory.java#L67
checkout this and it seems like they have a similar exception system that uses GRPC status as the indicator, so we can retry on certain exceptions like how we handle grpc call.

File: dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java [hunk 4]
```diff
@@ -127,36 +124,41 @@ public static AlluxioEtcdClient getInstance(AlluxioConfiguration conf) {
...
     }
-  }
-
-  /**
-   * Disconnect.
-   * @throws IOException
-   */
-  public void disconnect() throws IOException {
-    close();
+    throw new UnavailableException(
```
Comment (jja725) [inline dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:158]: eventually we would like to throw runtime exception if we fail at the end

File: dora/shell/src/main/java/alluxio/cli/fsadmin/command/NodesCommand.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import alluxio.conf.AlluxioConfiguration;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+
+/**
+ * Command to get or administrate worker nodes status.
+ */
+public class NodesCommand extends AbstractFsAdminCommand {
```
Comment (jja725) [inline dora/shell/src/main/java/alluxio/cli/fsadmin/command/NodesCommand.java:25]: would be nice if we can have an integration test

**Review by jja725 — COMMENTED** [review]
Mostly LGTM, leave some minor comments

File: dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java [hunk 6]
```diff
@@ -109,6 +102,10 @@ public AlluxioEtcdClient(AlluxioConfiguration conf) {
     List<String> endpointsList = conf.getList(PropertyKey.ETCD_ENDPOINTS);
```
Comment (jenoudet) [inline dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:102]: If the `AlluxioEtcdClient` is meant to be used as a singleton then its constructor should not be public, right?

File: dora/shell/src/main/java/alluxio/cli/fsadmin/command/NodesCommand.java (hunk 5 above)
Comment (jenoudet) [reply dora/shell/src/main/java/alluxio/cli/fsadmin/command/NodesCommand.java:25]: The bug fix for the etcd client connection and a new CLI command should be two separate PRs.

File: dora/shell/src/main/java/alluxio/cli/fsadmin/nodes/WorkerStatusCommand.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+ */
+public class WorkerStatusCommand extends AbstractFsAdminCommand {
+
+  private final AlluxioConfiguration mAlluxioConf;
+
+  /**
+   * @param context
+   * @param alluxioConf
+   */
+  public WorkerStatusCommand(Context context, AlluxioConfiguration alluxioConf) {
```
Comment (jenoudet) [inline dora/shell/src/main/java/alluxio/cli/fsadmin/nodes/WorkerStatusCommand.java:34]: Ditto, this should be in a separate PR.

File: dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java (hunk 4 above)
Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:158]: yeah I'm bit fuzzy on what exception to throw here,  so just referring to the exact layout in AbstractClient.java, let me talk to u offline separately

File: dora/shell/src/main/java/alluxio/cli/fsadmin/command/NodesCommand.java (hunk 5 above)
Comment (lucyge2022) [reply dora/shell/src/main/java/alluxio/cli/fsadmin/command/NodesCommand.java:25]: ok let me carve out this cmd piece and add integration test over there

File: dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java (hunk 6 above)
Comment (lucyge2022) [reply dora/core/common/src/main/java/alluxio/membership/AlluxioEtcdClient.java:102]: Yeah it was used in testcontainers UT before, let me add the @VisibleForTesting annotation.

File: dora/shell/src/main/java/alluxio/cli/fsadmin/nodes/WorkerStatusCommand.java (hunk 7 above)
Comment (lucyge2022) [reply dora/shell/src/main/java/alluxio/cli/fsadmin/nodes/WorkerStatusCommand.java]: done

**Review by jja725 — APPROVED** [review]
LGTM

**Comment by lucyge2022** [comment]
alluxio-bot, merge this please.

**Comment by lucyge2022** [comment]
alluxio-bot, merge this please.
